cmd/gardenlet/app/bootstrap.go

```diff
@@ -21,121 +21,65 @@ import (
 	"github.com/gardener/gardener/pkg/client/kubernetes"
 	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
 	"github.com/gardener/gardener/pkg/gardenlet/bootstrap"
-	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	bootstraputil "github.com/gardener/gardener/pkg/gardenlet/bootstrap/util"
 
 	"github.com/sirupsen/logrus"
-	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	certificatesv1beta1client "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
-	"k8s.io/client-go/tools/clientcmd"
-	"k8s.io/client-go/util/keyutil"
 	componentbaseconfig "k8s.io/component-base/config"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
+// bootstrapKubeconfig retrieves an already existing kubeconfig for the Garden cluster from the Seed or bootstraps a new one
 func bootstrapKubeconfig(
 	ctx context.Context,
 	logger *logrus.Logger,
-	gardenClientConnection *config.GardenClientConnection,
+	gardenClientConnection config.GardenClientConnection,
 	seedClientConnection componentbaseconfig.ClientConnectionConfiguration,
 	seedConfig *config.SeedConfig,
 ) (
 	[]byte,
-	string,
-	string,
+	*string,
+	*string,
```

Comment:
Why is this changed?

Reply:
To have a proper nil check [here](https://github.com/gardener/gardener/blob/8e40be5573473657c6cae5d01491649f32afaf94/cmd/gardenlet/app/gardenlet.go#L316) other than that it is identical to how it was.

Reply:
Hmm, you can check for an "unset" string by either checking its length (`len(s) == 0`) or checking equality with its zero value (`s == ""`). So why do you need a pointer here?

Reply:
Is returning a string rather than a pointer a cleaner design in your opinion? If yes, why? Does it have any other benefits?
Just trying to understand why you want to have the value returned.

Reply:
I think, passing an empty string and leveraging zero-values in general is the idiomatic way to do such things in go.
A clear benefit in the string case is, that you don't have to dereference the pointer. Also, you can directly return literals without assigning them to a variable first and then taking a reference...
Apart from that, what would be the semantic difference between `nil` and `&""`?
I think its a perfectly clear design to pass zero-values as an indicator that a value is "not present", so you don't need a pointer for that only to introduce yet another distinction between set and unset.

Reply:
Ok, makes sense to me. Thanks for the explanation.
Changed it.